src/System.Private.CoreLib/shared/System/Threading/Mutex.cs

```diff
@@ -0,0 +1,218 @@
+                    throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name, WaitHandleNameMax), nameof(name));
+#endif
+                if (errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+#if CORECLR
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+#else
+                throw ExceptionFromCreationError(errorCode, name);
+#endif
+            }
+
+            createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
+            SafeWaitHandle = mutexHandle;
+        }
+
+        private static OpenExistingResult OpenExistingWorker(string name, out Mutex result)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (name.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyName, nameof(name));
+            }
+
+            VerifyNameForCreate(name);
+            result = null;
+            // To allow users to view & edit the ACL's, call OpenMutex
+            // with parameters to allow us to view & edit the ACL.  This will
+            // fail if we don't have permission to view or edit the ACL's.  
+            // If that happens, ask for less permissions.
+#if CORECLR            
+            SafeWaitHandle myHandle = Win32Native.OpenMutex(AccessRights, false, name);
+#else
+            SafeWaitHandle myHandle = Interop.mincore.OpenMutex(AccessRights, false, name);
+#endif
+            if (myHandle.IsInvalid)
+            {
+                int errorCode = Marshal.GetLastWin32Error();
+
+#if PLATFORM_UNIX
+                if (name != null && errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
+                {
+                    // On Unix, length validation is done by CoreCLR's PAL after converting to utf-8
+                    throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name, WaitHandleNameMax), nameof(name));
+                }
+#endif
+                if (Interop.Errors.ERROR_FILE_NOT_FOUND == errorCode || Interop.Errors.ERROR_INVALID_NAME == errorCode)
+                    return OpenExistingResult.NameNotFound;
+                if (Interop.Errors.ERROR_PATH_NOT_FOUND == errorCode)
+                    return OpenExistingResult.PathNotFound;
+                if (null != name && Interop.Errors.ERROR_INVALID_HANDLE == errorCode)
+                    return OpenExistingResult.NameInvalid;
+
+#if CORECLR
+                // this is for passed through Win32Native Errors
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+#else
+                throw ExceptionFromCreationError(errorCode, name);
+#endif
+            }
+
+            result = new Mutex(myHandle);
+            return OpenExistingResult.Success;
+        }
+
+#if !CORECLR
```

Comment:
Why not keep the Windows specific implementation in the `.Windows.cs` file like it is in CoreRT now instead of these ifdefs?

Reply:
Because in Coreclr, the windows and unix implementation is almost the same

Reply:
That's because of it runs on Windows API emulator. We still call this Windows and add it here: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems#L785

Reply:
(PAL, FeaturePal, etc. is the Windows API emulator from CoreCLR.)

Reply:
Sorry I didn't Understand what can we do to reduce the ifdefs here ?

Reply:
- Keep the `.Windows.cs` file as it exists in CoreRT
- Add it to this block: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems#L785
- Rename `mincore` interop files to `Kernel32` to make it fit the naming conventions
- Work though build errors on Windows & Unix; add ifdefs for the differences - it should be a lot less than what you got now.